src/contracts/fungible/cache/sql.rs

```diff
@@ -264,6 +264,86 @@ impl Cache for SqlCache {
         self.save()?;
         Ok(existed)
     }
+
+    fn utxo_allocation_map(
+        &self,
+        contract_id: ContractId,
+    ) -> Result<HashMap<bitcoin::OutPoint, Vec<AtomicValue>>, CacheError> {
+        // Get the required asset from cache
+        let asset = self.asset(contract_id)?;
+
+        // Fetch its known_allocation map
+        let allocation_map = asset.known_allocations();
+
+        // Process known_allocation map to produce the intended map
+        let mut result = HashMap::new();
+
+        for item in allocation_map {
+            let mut alloc_amount = vec![];
+            for alloc in item.1 {
+                alloc_amount.push(alloc.value().value);
+            }
+            result.insert(item.0.clone(), alloc_amount);
+        }
+
+        Ok(result)
+    }
+
+    fn asset_allocation_map(
+        &self,
+        utxo: &bitcoin::OutPoint,
+    ) -> Result<HashMap<ContractId, Vec<AtomicValue>>, CacheError> {
+        // Explicitly local import
+        // Will cause name clash in global scope otherwise
+        use cache_schema::sql_allocation_utxo::dsl::*;
+
+        // fetch all the utxo data from sqlite table matching the txid and vout
+        // There can be multiple utxos having same (txid,vout) pair but linked
+        // with different assets
+        let sql_utxos = sql_allocation_utxo
+            .filter(txid.eq(utxo.txid.to_hex()))
+            .filter(vout.eq(utxo.vout as i32))
+            .load::<SqlAllocationUtxo>(&self.connection)
+            .map_err(|e| SqlCacheError::Sqlite(e))?;
+
+        // fetch all allocation data corresponding to the utxos
+        // and group them by their linked utxo
+        let allocations = SqlAllocation::belonging_to(&sql_utxos)
+            .load::<SqlAllocation>(&self.connection)
+            .map_err(|e| SqlCacheError::Sqlite(e))?
+            .grouped_by(&sql_utxos);
+
+        // Create a vector group of Vec<(Utxo, Vec<Allocation>)> for easier
+        // processing
+        let utxo_allocation_groups =
+            sql_utxos.into_iter().zip(allocations).collect::<Vec<_>>();
+
+        // Create the empty result map
+        let mut result = HashMap::new();
+
+        // Process the above groups to produce the required map
+        for item in utxo_allocation_groups {
+            let contract_id_string = sql_asset_table
+                .find(item.0.sql_asset_id)
+                .first::<SqlAsset>(&self.connection)
+                .map_err(|e| SqlCacheError::Sqlite(e))?
+                .contract_id
+                .clone();
+
+            let contract_id =
+                ContractId::from_hex(&contract_id_string[..]).unwrap();
+
+            let sum: Vec<AtomicValue> = item
+                .1
+                .iter()
+                .map(|alloc| alloc.amount as AtomicValue)
+                .collect();
+
+            result.insert(contract_id, sum);
+        }
```

Comment:
It would be nice to use instead `item` just `(contract_id, amount)` and get rid of all those hard to read `.0` and `.1`s

Reply:
fixed